oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/JdbcOAuth2AuthorizedClientService.java

```diff
@@ -0,0 +1,268 @@
+		Assert.hasText(principalName, "principalName cannot be empty");
+
+		Object[] args = {clientRegistrationId, principalName};
+		int[] argTypes = {Types.VARCHAR, Types.VARCHAR};
+		PreparedStatementSetter pss = new ArgumentTypePreparedStatementSetter(args, argTypes);
+
+		List<OAuth2AuthorizedClient> result = this.jdbcTemplate.query(
+				LOAD_AUTHORIZED_CLIENT_SQL,
+				pss,
+				(rs, rowNum) -> {
+					OAuth2AuthorizedClientData authorizedClientData = mapData(rs);
+
+					ClientRegistration clientRegistration = this.clientRegistrationRepository.findByRegistrationId(
+							authorizedClientData.clientRegistrationId);
+					if (clientRegistration == null) {
+						throw new DataRetrievalFailureException("The ClientRegistration with id '" +
+								authorizedClientData.clientRegistrationId + "' exists in the data source, " +
+								"however, it was not found in the ClientRegistrationRepository.");
+					}
+
+					OAuth2AccessToken.TokenType tokenType = null;
+					if (OAuth2AccessToken.TokenType.BEARER.getValue().equalsIgnoreCase(authorizedClientData.accessTokenType)) {
+						tokenType = OAuth2AccessToken.TokenType.BEARER;
+					}
+					String tokenValue = authorizedClientData.accessTokenValue;		// TODO Decrypt
+					Instant issuedAt = authorizedClientData.accessTokenIssuedAt.toInstant();
+					Instant expiresAt = authorizedClientData.accessTokenExpiresAt.toInstant();
+					Set<String> scopes = Collections.emptySet();
+					if (authorizedClientData.accessTokenScopes != null) {
+						scopes = StringUtils.commaDelimitedListToSet(authorizedClientData.accessTokenScopes);
+					}
+					OAuth2AccessToken accessToken = new OAuth2AccessToken(
+							tokenType, tokenValue, issuedAt, expiresAt, scopes);
+
+					OAuth2RefreshToken refreshToken = null;
+					if (authorizedClientData.refreshTokenValue != null) {
+						tokenValue = authorizedClientData.refreshTokenValue;	// TODO Decrypt
+						issuedAt = null;
+						if (authorizedClientData.refreshTokenIssuedAt != null) {
+							issuedAt = authorizedClientData.refreshTokenIssuedAt.toInstant();
+						}
+						refreshToken = new OAuth2RefreshToken(tokenValue, issuedAt);
+					}
+
+					return new OAuth2AuthorizedClient(clientRegistration,
+							authorizedClientData.principalName, accessToken, refreshToken);
+				});
+
+		return !result.isEmpty() ? (T) result.get(0) : null;
+	}
+
+	@Transactional
+	@Override
+	public void saveAuthorizedClient(OAuth2AuthorizedClient authorizedClient, Authentication principal) {
+		Assert.notNull(authorizedClient, "authorizedClient cannot be null");
+		Assert.notNull(principal, "principal cannot be null");
+
+		OAuth2AuthorizedClientData authorizedClientData = mapData(authorizedClient, principal);
+		Object[] args = {
+				authorizedClientData.clientRegistrationId,
+				authorizedClientData.principalName,
+				authorizedClientData.accessTokenType,
+				authorizedClientData.accessTokenValue,		// TODO Encrypt
```

Comment:
@rwinch Do you have a preference/recommendation on the approach for encrypting/securing the access token and refresh token?

Reply:
I think we should externalize the mapping (reading/writing) of the data into one of Spring's strategies. Then user's can inject their own strategies.

Reply:
Which specific strategy were you thinking? `Converter` or you had another one in mind?